File: packages/shared/lib/src/models/shipment.dart
import '../enums/shipment_status.dart';
import '../enums/service_type.dart';

class ShipmentItem {
  final String? id;
  final String description;
  final int quantity;
  final double weight;
  final double? length;
  final double? width;
  final double? height;
  final double? volumeWeight;

  ShipmentItem({
    this.id,
    required this.description,
    required this.quantity,
    required this.weight,
    this.length,
    this.width,
    this.height,
    this.volumeWeight,
  });

  factory ShipmentItem.fromJson(Map<String, dynamic> json) {
    return ShipmentItem(
      id: json['id'] as String?,
      description: json['description'] as String,
      quantity: json['quantity'] as int,
      weight: (json['weight'] as num).toDouble(),
      length: (json['length'] as num?)?.toDouble(),
      width: (json['width'] as num?)?.toDouble(),
      height: (json['height'] as num?)?.toDouble(),
      volumeWeight: (json['volume_weight'] as num?)?.toDouble(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'description': description,
      'quantity': quantity,
      'weight': weight,
      'length': length,
      'width': width,
      'height': height,
      'volume_weight': volumeWeight,
    };
  }
}

class Shipment {
  final String id;
  final String trackingNumber;
  final ServiceType serviceType;
  final ShipmentStatus status;
  final String? customerId;
  final String? driverId;
  final String? vehicleId;
  final String? containerId;
  final String? voyageId;
  final String? branchId;
  final String pickupAddress;
  final String deliveryAddress;
  final double? pickupLatitude;
  final double? pickupLongitude;
  final double? deliveryLatitude;
  final double? deliveryLongitude;
  final String? senderName;
  final String? senderPhone;
  final String? recipientName;
  final String? recipientPhone;
  final double? estimatedPrice;
  final double? finalPrice;
  final String? notes;
  final List<ShipmentItem> items;
  final DateTime createdAt;
  final DateTime updatedAt;

  Shipment({
    required this.id,
    required this.trackingNumber,
    required this.serviceType,
    required this.status,
    this.customerId,
    this.driverId,
    this.vehicleId,
    this.containerId,
    this.voyageId,
    this.branchId,
    required this.pickupAddress,
    required this.deliveryAddress,
    this.pickupLatitude,
    this.pickupLongitude,
    this.deliveryLatitude,
    this.deliveryLongitude,
    this.senderName,
    this.senderPhone,
    this.recipientName,
    this.recipientPhone,
    this.estimatedPrice,
    this.finalPrice,
    this.notes,
    this.items = const [],
    required this.createdAt,
    required this.updatedAt,
  });

  factory Shipment.fromJson(Map<String, dynamic> json) {
    return Shipment(
      id: json['id'] as String,
      trackingNumber: json['tracking_number'] as String,
      serviceType: ServiceType.fromApi(json['service_type'] as String),
      status: ShipmentStatus.fromApi(json['status'] as String),
      customerId: json['customer_id'] as String?,
      driverId: json['driver_id'] as String?,
      vehicleId: json['vehicle_id'] as String?,
      containerId: json['container_id'] as String?,
      voyageId: json['voyage_id'] as String?,
      branchId: json['branch_id'] as String?,
      pickupAddress: json['pickup_address'] as String,
      deliveryAddress: json['delivery_address'] as String,
      pickupLatitude: (json['pickup_latitude'] as num?)?.toDouble(),
      pickupLongitude: (json['pickup_longitude'] as num?)?.toDouble(),
      deliveryLatitude: (json['delivery_latitude'] as num?)?.toDouble(),
      deliveryLongitude: (json['delivery_longitude'] as num?)?.toDouble(),
      senderName: json['sender_name'] as String?,
      senderPhone: json['sender_phone'] as String?,
      recipientName: json['recipient_name'] as String?,
      recipientPhone: json['recipient_phone'] as String?,
      estimatedPrice: (json['estimated_price'] as num?)?.toDouble(),
      finalPrice: (json['final_price'] as num?)?.toDouble(),
      notes: json['notes'] as String?,
      items: (json['items'] as List<dynamic>?)
              ?.map((e) => ShipmentItem.fromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'tracking_number': trackingNumber,
      'service_type': serviceType.apiValue,
      'status': status.apiValue,
      'customer_id': customerId,
      'driver_id': driverId,
      'vehicle_id': vehicleId,
      'container_id': containerId,
      'voyage_id': voyageId,
      'branch_id': branchId,
      'pickup_address': pickupAddress,
      'delivery_address': deliveryAddress,
      'pickup_latitude': pickupLatitude,
      'pickup_longitude': pickupLongitude,
      'delivery_latitude': deliveryLatitude,
      'delivery_longitude': deliveryLongitude,
      'sender_name': senderName,
      'sender_phone': senderPhone,
      'recipient_name': recipientName,
      'recipient_phone': recipientPhone,
      'estimated_price': estimatedPrice,
      'final_price': finalPrice,
      'notes': notes,
      'items': items.map((e) => e.toJson()).toList(),
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
    };
  }
}

File: backend/api/app/models/shipment.py
import enum
import uuid
from datetime import datetime

from sqlalchemy import DateTime, Enum, Float, ForeignKey, Integer, String, Text, func
from app.database import GUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base


class ServiceType(str, enum.Enum):
    international_road = "international_road"
    maritime = "maritime"
    domestic = "domestic"


class ShipmentStatus(str, enum.Enum):
    pending = "pending"
    confirmed = "confirmed"
    picked_up = "picked_up"
    in_transit = "in_transit"
    at_terminal = "at_terminal"
    customs_clearance = "customs_clearance"
    out_for_delivery = "out_for_delivery"
    delivered = "delivered"
    failed_attempt = "failed_attempt"
    delayed = "delayed"
    on_hold = "on_hold"
    cancelled = "cancelled"
    returned = "returned"


class Shipment(Base):
    __tablename__ = "shipments"

    id: Mapped[uuid.UUID] = mapped_column(GUID(), primary_key=True, default=uuid.uuid4)
    tracking_number: Mapped[str] = mapped_column(String(50), unique=True, nullable=False, index=True)
    service_type: Mapped[ServiceType] = mapped_column(Enum(ServiceType), nullable=False)
    status: Mapped[ShipmentStatus] = mapped_column(Enum(ShipmentStatus), default=ShipmentStatus.pending, nullable=False)
    customer_id: Mapped[uuid.UUID] = mapped_column(GUID(), ForeignKey("customers.id"), nullable=False)
    driver_id: Mapped[uuid.UUID | None] = mapped_column(GUID(), ForeignKey("drivers.id"), nullable=True)
    vehicle_id: Mapped[uuid.UUID | None] = mapped_column(GUID(), ForeignKey("vehicles.id"), nullable=True)
    container_id: Mapped[uuid.UUID | None] = mapped_column(GUID(), ForeignKey("containers.id"), nullable=True)
    voyage_id: Mapped[uuid.UUID | None] = mapped_column(GUID(), ForeignKey("voyages.id"), nullable=True)
    branch_id: Mapped[uuid.UUID | None] = mapped_column(GUID(), ForeignKey("branches.id"), nullable=True)
    pickup_address: Mapped[str | None] = mapped_column(Text, nullable=True)
    delivery_address: Mapped[str | None] = mapped_column(Text, nullable=True)
    pickup_latitude: Mapped[float | None] = mapped_column(Float, nullable=True)
    pickup_longitude: Mapped[float | None] = mapped_column(Float, nullable=True)
    delivery_latitude: Mapped[float | None] = mapped_column(Float, nullable=True)
    delivery_longitude: Mapped[float | None] = mapped_column(Float, nullable=True)
    sender_name: Mapped[str] = mapped_column(String(255), nullable=False)
    sender_phone: Mapped[str] = mapped_column(String(20), nullable=False)
    recipient_name: Mapped[str] = mapped_column(String(255), nullable=False)
    recipient_phone: Mapped[str] = mapped_column(String(20), nullable=False)
    estimated_price: Mapped[float | None] = mapped_column(Float, nullable=True)
    final_price: Mapped[float | None] = mapped_column(Float, nullable=True)
    weight: Mapped[float | None] = mapped_column(Float, nullable=True)
    volume_weight: Mapped[float | None] = mapped_column(Float, nullable=True)
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())

    items = relationship("ShipmentItem", back_populates="shipment", cascade="all, delete-orphan")
    tracking_events = relationship("TrackingEvent", back_populates="shipment", cascade="all, delete-orphan")
    documents = relationship("Document", back_populates="shipment", cascade="all, delete-orphan")


class ShipmentItem(Base):
    __tablename__ = "shipment_items"

    id: Mapped[uuid.UUID] = mapped_column(GUID(), primary_key=True, default=uuid.uuid4)
    shipment_id: Mapped[uuid.UUID] = mapped_column(GUID(), ForeignKey("shipments.id"), nullable=False)
    description: Mapped[str] = mapped_column(Text, nullable=False)
    quantity: Mapped[int] = mapped_column(Integer, default=1)
    weight: Mapped[float | None] = mapped_column(Float, nullable=True)
    length: Mapped[float | None] = mapped_column(Float, nullable=True)
    width: Mapped[float | None] = mapped_column(Float, nullable=True)
    height: Mapped[float | None] = mapped_column(Float, nullable=True)
    volume_weight: Mapped[float | None] = mapped_column(Float, nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())

    shipment = relationship("Shipment", back_populates="items")
